src/GitHub.TeamFoundation.14/Services/VSGitExt.cs

```diff
@@ -1,39 +1,90 @@
 using System;
+using System.Linq;
 using System.Collections.Generic;
 using System.ComponentModel.Composition;
-using System.Linq;
-using GitHub.Extensions;
 using GitHub.Models;
 using GitHub.Services;
+using GitHub.Logging;
+using Serilog;
 using Microsoft.VisualStudio.TeamFoundation.Git.Extensibility;
 
 namespace GitHub.VisualStudio.Base
 {
+    /// <summary>
+    /// This service acts as an always available version of <see cref="IGitExt"/>.
+    /// </summary>
     [Export(typeof(IVSGitExt))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     public class VSGitExt : IVSGitExt
     {
+        static readonly ILogger log = LogManager.ForContext<VSGitExt>();
+
+        IGitHubServiceProvider serviceProvider;
+        IVSUIContext context;
         IGitExt gitService;
 
-        public void Refresh(IServiceProvider serviceProvider)
+        [ImportingConstructor]
+        public VSGitExt(IGitHubServiceProvider serviceProvider, IVSUIContextFactory factory)
         {
-            if (gitService != null)
-                gitService.PropertyChanged -= CheckAndUpdate;
-            gitService = serviceProvider.GetServiceSafe<IGitExt>();
-            if (gitService != null)
-                gitService.PropertyChanged += CheckAndUpdate;
+            this.serviceProvider = serviceProvider;
+
+            // The IGitExt service is only available when in the SccProvider context.
+            // This could be changed to VSConstants.UICONTEXT.SolutionExists_guid when testing.
+            context = factory.GetUIContext(new Guid(Guids.GitSccProviderId));
+
+            // If we're not in the GitSccProvider context or TryInitialize fails, have another go when the context changes.
+            if (!context.IsActive || !TryInitialize())
+            {
+                context.UIContextChanged += ContextChanged;
+                log.Information("VSGitExt will be initialized later");
+            }
         }
 
+        void ContextChanged(object sender, VSUIContextChangedEventArgs e)
+        {
+            // If we're in the GitSccProvider context and TryInitialize succeeds, we can stop listening for events.
```

Comment:
We also need to listen for the context to be deactivated, and throw away the instance of GitExt if that happens, and get it back when it activates again.

Reply:
I've done a bunch of experiments changing to and from solutions in TFS repos.  Here are some findings:

- When changing to a TFS repo there can be two `UIContext` events, first with `IsActive=false` then with `IsActive=true`
- When changing back to a Git repo there are (often?) no events fired
- The `IGitExt` service will only be null when directly opening a TFS solution
- Once set the returned `IGitExt` service object will remain constant not return

This is consistent with the `Microsoft.TeamFoundation.Git.Provider.SccExtensibilityService` being delay loaded but never unloaded.

Given that the context is deactivated and then activated again when opening a TFS based solution, I'm reluctant to do anything to the `IGitExt` object when this happens. We'd still be listening when in a TFS solution and continue to use the same object when returning to a Git based solution.

Does that make sense? I should update the comments to reflect this.